docs/source/argument_check.rst

```diff
@@ -0,0 +1,636 @@
+
+- For vectors and matrices with batched stride, there is no argument checking for stride. To access elements in a strided_batched_matrix, for example the C matrix in gemm, the zero based index is calculated as ``i1 + i2 * ldc + i3 * stride_c``, where ``i1 = 0, 1, 2, … m-1``; ``i2 = 0, 1, 2,  … n-1``; ``i3 = 0, 1, 2, … batch_count -1``. An incorrect stride can result in a core dump due a segmentation fault. It can also produce an indeterminate result if there is a memory overlap in the output matrix between different values of ``i3``.
+
+
+Device Memory Size Queries
+==========================
+
+- When ``handle->is_device_memory_size_query()`` is true, the call is not a normal call, but it is a device memory size query.
+
+- No logging should be performed during device memory size queries.
+
+- If the rocBLAS kernel requires no temporary device memory, the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle)`` can be called after checking that ``handle != nullptr``.
+
+- If the rocBLAS kernel requires temporary device memory, then it should be set, and the kernel returned, by calling ``return handle->set_optimal_device_memory_size(size...)``, where ``size...`` is a list of one or more sizes for different sub-problems. The sizes are rounded up and added.
+
+Logging
+-------
+
+- There is logging before a quick-return-success or error-return, except:
+
+  - when ``handle == nullptr``, return ``rocblas_status_invalid_handle`` 
+  - when ``handle->is_device_memory_size_query()`` returns ``true``
+
+- Vectors and matrices are logged with their addresses, and are always on device memory.
+
+- Scalar values in device memory are logged as their addresses. Scalar values in host memory are logged as their values, with a ``nullptr`` logged as ``NaN`` (``std::numeric_limits<T>::quiet_NaN()``).
+
+rocBLAS control flow:
+=====================
+
+1. If ``handle == nullptr``, return ``rocblas_status_invalid_handle``.
+
+2. If the function does not require temporary device memory, call the macro ``RETURN_ZERO_DEVICE_MEMORY_SIZE_IF_QUERIED(handle);``.
+
+3. If the function requires temporary device memory, and ``handle->is_device_memory_size_query()`` is ``true``, then validate any pointers and arguments required to determine the optimal size of temporary device memory, returning ``rocblas_status_invalid_pointer`` or ``rocblas_status_invalid_size`` if the arguments are invalid, and otherwise ``return handle->set_optimal_device_memory_size(size...);``, where ``size...`` is a list of one or more sizes of temporary buffers, which are allocated with ``handle->device_malloc(size...)`` later.
+
+4. Perform logging if enabled, taking care not to dereference ``nullptr`` arguments.
+
+5. Check for invalid sizes. Return ``rocblas_status_invalid_size`` if size arguments are invalid.
```

Comment:
5 and 6 might be backwards, I think in LAPACK they check enums (rather, characters, for them) first then invalid_sizes.